File: cs/hsocket/hclient.cs
﻿using System.Net;
using System.Net.Sockets;
using System.Text.Json.Nodes;

namespace hsocket;


enum BuiltInOpCode : ushort
{
    FT_TRANSFER_PORT = 60020,  // 文件传输端口{ "port": port }
    FT_SEND_FILES_HEADER = 62000
}


public abstract class HTcpClient
{
    public delegate void OnConnectedCallback();
    public delegate void OnDisconnectedCallback();

    public event OnConnectedCallback? OnConnectedEvent;
    public event OnDisconnectedCallback? OnDisconnectedEvent;

    protected HTcpSocket tcpSocket = new();
    protected IPAddress? ftServerIp = null;
    protected int ftServerPort = 0;

    public HTcpClient()
    {
        tcpSocket.Blocking = true;
    }

    public HTcpSocket Socket()
    {
        return tcpSocket;
    }

    public virtual void Connect(IPEndPoint addr)
    {
        tcpSocket.Connect(addr);
        ftServerIp = addr.Address;
        OnConnected();
    }

    public virtual void Close()
    {
        tcpSocket.Close();
    }

    public virtual bool IsClosed()
    {
        return !tcpSocket.IsValid();
    }

    /// <summary>
    /// 向服务端发送报文
    /// </summary>
    /// <param name="msg">发送的报文</param>
    /// <returns>是否成功</returns>
    public abstract bool SendMsg(Message msg);

    /// <summary>
    /// 获取服务端的文件传输端口
    /// </summary>
    /// <returns>是否成功获取</returns>
    protected abstract bool GetFTTransferPort();

    /// <summary>
    /// 发送一个文件
    /// </summary>
    /// <param name="path">文件路径</param>
    /// <param name="filename">文件名</param>
    /// <exception cref="NullReferenceException"></exception>
    public void SendFile(string path, string filename)
    {
        if (ftServerIp == null)
            throw new NullReferenceException("ftServerIp is null");
        if (!GetFTTransferPort())
            return;
        // send
        using HTcpSocket ftSocket = new();
        FileStream? fin = null;
        try
        {
            ftSocket.Connect(new IPEndPoint(ftServerIp, ftServerPort));
            try
            {
                fin = File.OpenRead(path);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
                return;
            }
            ftSocket.SendFile(fin, filename);
        }
        catch
        {
            return;
        }
        finally
        {
            fin?.Close();
        }
    }

    /// <summary>
    /// 接收一个文件
    /// </summary>
    /// <returns>下载的文件路径，失败时返回空字符串</returns>
    /// <exception cref="NullReferenceException"></exception>
    public string RecvFile()
    {
        if (ftServerIp == null)
            throw new NullReferenceException("'ftServerIp' is null.");
        if (!GetFTTransferPort())
            return "";
        // recv
        using HTcpSocket ftSocket = new();
        try
        {
            ftSocket.Connect(new IPEndPoint(ftServerIp, ftServerPort));
            string downloadPath = ftSocket.RecvFile();
            return downloadPath;
        }
        catch
        {
            return "";
        }
    }

    /// <summary>
    /// 发送多个文件
    /// </summary>
    /// <param name="paths">文件路径列表</param>
    /// <param name="filenames">文件名列表</param>
    /// <returns>成功发送的文件数</returns>
    /// <exception cref="NullReferenceException"></exception>
    /// <exception cref="ArgumentException"></exception>
    public List<string> SendFiles(List<string> paths, List<string> filenames)
    {
        if (ftServerIp == null)
            throw new NullReferenceException("'ftServerIp' is null.");
        if (!GetFTTransferPort())
            return new();
        // send
        List<string> succeedPathList = new();
        using HTcpSocket ftSocket = new();
        try
        {
            ftSocket.Connect(new IPEndPoint(ftServerIp, ftServerPort));
            ftSocket.SendFiles(paths, filenames, ref succeedPathList);
        }
        catch (ArgumentException) { throw; }
        catch { }
        return succeedPathList;
    }

    /// <summary>
    /// 接收多个文件
    /// </summary>
    /// <returns>下载的文件路径列表</returns>
    /// <exception cref="NullReferenceException"></exception>
    public List<string> RecvFiles()
    {
        if (ftServerIp == null)
            throw new NullReferenceException("'ftServerIp' is null.");
        if (!GetFTTransferPort())
            return new List<string>();
        // recv
        List<string> downloadPathList = new();
        try
        {
            using HTcpSocket ftSocket = new();
            ftSocket.Connect(new IPEndPoint(ftServerIp, ftServerPort));
            ftSocket.RecvFiles(ref downloadPathList);
        }
        catch { }
        return downloadPathList;
    }

    protected void OnConnected()
    {
        OnConnectedEvent?.Invoke();
    }
    protected void OnDisconnected()
    {
        OnDisconnectedEvent?.Invoke();
    }
}

File: cpp/src/hsocket/source/hclient.cpp
#include "pch.h"
#include "hclient.h"
#include <fstream>
#include <iostream>

HTcpClient::HTcpClient()
{
	tcpSocket.setblocking(true);
}

void HTcpClient::connect(IPv4Address addr)
{
	tcpSocket.connect(addr.ip.c_str(), addr.port);
	ftServerIp = addr.ip;
	this->onConnected();
}

void HTcpClient::close()
{
	tcpSocket.close();
}

void HTcpClient::sendfile(const std::string &path, const std::string &filename)
{
	if (!this->getFTTransferPort())
		return;
	// send
	HTcpSocket ftSocket;
	try {
		ftSocket.connect(ftServerIp.c_str(), ftServerPort);
	}
	catch (SocketError e) {
		ftSocket.close();
		return;
	}
	std::ifstream fin = std::ifstream(path, std::ios::binary | std::ios::in);
	if (fin.fail()) {
		std::cout << "File not found: " << "\"" << path << "\"" << std::endl;
		return;
	}
	try {
		ftSocket.sendFile(fin, filename);
	}
	catch (SocketError e) {}
	fin.close();
	ftSocket.close();
}

std::string HTcpClient::recvfile()
{
	if (!this->getFTTransferPort())
		return "";
	// recv
	HTcpSocket ftSocket;
	std::string downPath = "";
	try {
		ftSocket.connect(ftServerIp.c_str(), ftServerPort);
		downPath = ftSocket.recvFile();
	}
	catch (SocketError e) {}
	ftSocket.close();
	return downPath;
}

std::vector<std::string> HTcpClient::sendfiles(std::vector<std::string> &paths, std::vector<std::string> &filenames)
{
	if (!this->getFTTransferPort())
		return {};
	// send
	std::vector<std::string> succeedPathList;
	HTcpSocket ftSocket;
	try {
		ftSocket.connect(ftServerIp.c_str(), ftServerPort);
		ftSocket.sendFiles(paths, filenames, succeedPathList);
	}
	catch (std::invalid_argument e) {
		return {};
	}
	catch (SocketError e) {	}
	ftSocket.close();
	return succeedPathList;
}

std::vector<std::string> HTcpClient::recvfiles()
{
	if (!this->getFTTransferPort())
		return std::vector<std::string>();
	// recv
	std::vector<std::string> downloadPathList;
	HTcpSocket ftSocket;
	try {
		ftSocket.connect(ftServerIp.c_str(), ftServerPort);
		ftSocket.recvFiles(downloadPathList);
	}
	catch (SocketError e) {}
	catch (MessageError e) {}
	ftSocket.close();
	return downloadPathList;
}


HTcpChannelClient::HTcpChannelClient()
	: HTcpClient()
{
}

void HTcpChannelClient::connect(IPv4Address addr)
{
	HTcpClient::connect(addr);
	thMessage = std::thread(&HTcpChannelClient::messageHandle, this);
}

bool HTcpChannelClient::sendmsg(const Message &msg)
{
	try {
		tcpSocket.sendMsg(msg);
	}
	catch (SocketError e) {
		thMessage.join();  // make sure that 'onDisconnected' only runs once
		if (!isClosed()) {
			this->onDisconnected();
			this->close();
		}
		return false;
	}
	return true;
}

void HTcpChannelClient::messageHandle()
{
	while (!this->isClosed()) {
		try {
			Message msg = tcpSocket.recvMsg();
			if (msg.opcode() == BuiltInOpCode::FT_TRANSFER_PORT) {
				ftServerPort = msg.getInt("port");
				conFTPort.notify_one();
				continue;
			}
			this->onMessageReceived(msg);
		}
		catch (SocketError e) {
			if (e.getErrcode() == EAGAIN || e.getErrcode() == EWOULDBLOCK) {
				continue;
			}
			this->onDisconnected();
			this->close();
			break;
		}
		catch (MessageError e) {
			this->onDisconnected();
			this->close();
			break;
		}
	}
}

bool HTcpChannelClient::getFTTransferPort()
{
	std::unique_lock<std::mutex> ulock(mtxFTPort);
	std::cv_status status = conFTPort.wait_for(ulock, std::chrono::milliseconds(ftTimeout));
	if (status == std::cv_status::no_timeout)
		return true;
	else
		return false;
}

void HTcpChannelClient::onMessageReceived(Message& msg)
{
	ushort opcode = msg.opcode();
	auto iter = onMsgRecvByOpCodeCallbackMap.find(opcode);
	if (iter != onMsgRecvByOpCodeCallbackMap.end()) {
		bool finished = iter->second(msg);
		if (finished)
			return;
	}
	if (onMessageReceivedCallback)
		onMessageReceivedCallback(msg);
}


HTcpReqResClient::HTcpReqResClient()
	: HTcpClient()
{
}

bool HTcpReqResClient::sendmsg(const Message &msg)
{
	try {
		tcpSocket.sendMsg(msg);
	}
	catch (SocketError e) {
		if (!isClosed()) {
			this->onDisconnected();
			this->close();
		}
		return false;
	}
	return true;
}

std::optional<Message> HTcpReqResClient::request(const Message &msg)
{
	if (this->sendmsg(msg)) {
		bool flagError = false;
		try {
			Message response = tcpSocket.recvMsg();
			return response;
		}
		catch (SocketError e) {
			flagError = true;
		}
		catch (MessageError e) {
			flagError = true;  // 收到空包文时断开
		}
		if (flagError) {
			this->onDisconnected();
			this->close();
			return std::nullopt;
		}
	}
	return std::nullopt;
}

bool HTcpReqResClient::getFTTransferPort()
{
	try {
		Message msg = tcpSocket.recvMsg();
		if (msg.opcode() == BuiltInOpCode::FT_TRANSFER_PORT) {
			ftServerPort = msg.getInt("port");
			return true;
		}
		else
			return false;
	}
	catch (SocketError e) {
		return false;
	}
	catch (MessageError e) {
		return false;
	}
}
